Tutorial on sparsifying BERT on SQuAD (#286)
neuralmagic/sparseml: integrations/huggingface-transformers/recipes/bert-base-6layers_prune95.md

<!--
Copyright (c) 2021 - present / Neuralmagic, Inc. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
-->

---

> # General Epoch/LR variables

Reviewer: ```suggestion
# General Epoch/LR Variables
```

num_epochs: &num_epochs 30

> # pruning hyperparameters

Reviewer: ```suggestion
# Pruning Hyperparameters
```

init_sparsity: &init_sparsity 0.00
final_sparsity: &final_sparsity 0.95
pruning_start_epoch: &pruning_start_epoch 2
pruning_end_epoch: &pruning_end_epoch 20
update_frequency: &pruning_update_frequency 0.01

> # modifiers:

Reviewer: ```suggestion
# Modifiers:
```

training_modifiers:
  - !EpochRangeModifier
    end_epoch: 30
    start_epoch: 0.0

pruning_modifiers:
  - !GMPruningModifier
    params:
      - re:bert.encoder.layer.([0,2,4,6,8]|11).attention.self.query.weight
      - re:bert.encoder.layer.([0,2,4,6,8]|11).attention.self.key.weight
      - re:bert.encoder.layer.([0,2,4,6,8]|11).attention.self.value.weight
      - re:bert.encoder.layer.([0,2,4,6,8]|11).attention.output.dense.weight
      - re:bert.encoder.layer.([0,2,4,6,8]|11).intermediate.dense.weight
      - re:bert.encoder.layer.([0,2,4,6,8]|11).output.dense.weight
    start_epoch: *pruning_start_epoch
    end_epoch: *pruning_end_epoch
    init_sparsity: *init_sparsity
    final_sparsity: *final_sparsity
    inter_func: cubic
    update_frequency: *pruning_update_frequency
    leave_enabled: True
    mask_type: unstructured
    log_types: __ALL__

  - !LayerPruningModifier
    layers:
      - bert.encoder.layer.1
      - bert.encoder.layer.3
      - bert.encoder.layer.5
      - bert.encoder.layer.7
      - bert.encoder.layer.9
      - bert.encoder.layer.10
---

> # Bert model with dropped and pruned encoder layers

Reviewer: ```suggestion
# BERT Model with Dropped and Pruned Encoder Layers
```

> This recipe defines a dropping and pruning strategy to sparsify 6 encoder layers of a Bert model at 95% sparsity. It was used together with knowledge distillation to create sparse model that achives 90% recovery from its (teacher) baseline accuracy on the Squad dataset. 

Reviewer: ```suggestion
This recipe defines a dropping and pruning strategy to sparsify six encoder layers of a BERT model at 95% sparsity. It was used together with knowledge distillation to create a sparse model that achieves 90% recovery from its (teacher) baseline accuracy on the SQuaD dataset. 
```

Training was done using 1 GPU at half precision using a training batch size of 16 with the
[SparseML integration with huggingface/transformers](https://github.com/neuralmagic/sparseml/tree/main/integrations/huggingface-transformers).

## Weights and Biases

> - [Sparse Bert on Squad](https://wandb.ai/neuralmagic/sparse-bert-squad/runs/3plynclw?workspace=user-neuralmagic)

Reviewer: ```suggestion
- [Sparse BERT on SQuAD](https://wandb.ai/neuralmagic/sparse-bert-squad/runs/3plynclw?workspace=user-neuralmagic)
```

-->

---
# General Variables
num_epochs: &num_epochs 30

# Pruning Hyperparameters
init_sparsity: &init_sparsity 0.00
final_sparsity: &final_sparsity 0.95
pruning_start_epoch: &pruning_start_epoch 2
pruning_end_epoch: &pruning_end_epoch 20
update_frequency: &pruning_update_frequency 0.01

> # Modifiers:

Reviewer: ```suggestion
# Modifiers
```

      - bert.encoder.layer.10
---

# BERT Model with Dropped and Pruned Encoder Layers

> This recipe defines a dropping and pruning strategy to sparsify six encoder layers of a BERT model at 95% sparsity. It was used together with knowledge distillation to create sparse model that achives 90% recovery from its (teacher) baseline accuracy on the SQuAD dataset. 

Reviewer: ```suggestion
This recipe defines a dropping and pruning strategy to sparsify six encoder layers of a BERT model at 95% sparsity. It was used together with knowledge distillation to create sparse model that achieves 90% recovery from its (teacher) baseline accuracy on the SQuAD dataset. 
```
